modules/rostests/apitests/user32/RedrawWindow.c

```diff
@@ -86,3 +91,1084 @@ START_TEST(RedrawWindow)
+    HWND hwnd = CreateWindowExW(0, wc.lpszClassName, wc.lpszClassName, WS_OVERLAPPEDWINDOW,
+        CW_USEDEFAULT, CW_USEDEFAULT, width, height, NULL, NULL, GetModuleHandle(NULL), NULL);
+    if (hwnd == NULL)
+        return;
+
+    ShowWindow(hwnd, SW_SHOW);
+    if(!ptestRW->testChild)
+        UpdateWindow(hwnd);
+
+    if (ptestRW->testChild)
+    {
+        WNDCLASSW wcChild = { };
+        wcChild.lpfnWndProc = ChildWindowProc;
+        wcChild.hInstance = GetModuleHandle(NULL);
+        wcChild.lpszClassName = CHILD_CLASS_NAME;
+        RegisterClassW(&wcChild);
+
+        hChildWnd = CreateWindowExW(
+            0,
+            CHILD_CLASS_NAME,
+            L"Child Window",
+            WS_CHILD | WS_VISIBLE,
+            10, 10, 200, 200,
+            hwnd,
+            NULL,
+            GetModuleHandle(NULL),
+            NULL
+        );
+    }
+
+    HDC hdc = GetDC(hwnd);
+    RECT drect = { 0, 0, 800, 600 };
+    DrawContent(hdc, &drect, RGB(255, 0, 0));
+    ReleaseDC(hwnd, hdc);
+
+    RgnUpdate = NULL;
+    if (ptestRW->useRegion) {
+        RgnUpdate = CreateRectRgn(ptestRW->regRect.left, ptestRW->regRect.top, ptestRW->regRect.right, ptestRW->regRect.bottom);
+    }
+
+    prect=NULL;
+    if (ptestRW->useRect) {
+        prect = &ptestRW->rectRect;
+    }
+
+    if (ptestRW->testChild)
+    {
+        ServeSomeMessages(10, 10);
+    }
+
+    ptestRW->resultRedraw = RedrawWindow(hwnd, prect, RgnUpdate, ptestRW->flags);
+
+    if (ptestRW->testChild)
+    {
+        ServeSomeMessages(10, 10);
+    }
+
+    hdc = GetDC(hwnd);
+    ptestRW->resultColorPre1 = GetPixel(hdc, ptestRW->testPixelPre1x, ptestRW->testPixelPre1y);
+    ptestRW->resultColorPre2 = GetPixel(hdc, ptestRW->testPixelPre2x, ptestRW->testPixelPre2y);
+    ReleaseDC(hwnd, hdc);
+
+    ptestRW->resultNeedsUpdate = GetUpdateRect(hwnd, &ptestRW->resultUpdateRect, FALSE);
+
+    if (ptestRW->forcePaint) {
+        UpdateWindow(hwnd);
+    }
+
+    hdc = GetDC(hwnd);
+    ptestRW->resultColorPost1 = GetPixel(hdc, ptestRW->testPixelPost1x, ptestRW->testPixelPost1y);
+    ptestRW->resultColorPost2 = GetPixel(hdc, ptestRW->testPixelPost2x, ptestRW->testPixelPost2y);
+    ReleaseDC(hwnd, hdc);
+
+    ptestRW->resultWmEraseGnd = resultWmEraseGnd;
+    ptestRW->resultWmNcPaint = resultWmNcPaint;
+    ptestRW->resultPaintIndex = paintIndex;
+
+    if (RgnUpdate) DeleteObject(RgnUpdate);
+
+    if (hChildWnd != NULL)
+        DestroyWindow(hChildWnd);
+    if (hwnd != NULL)
+        DestroyWindow(hwnd);
+}
+
+UINT TestRedrawWindow2(STRUCT_TestRedrawWindow* ptestRW, STRUCT_TestRedrawWindowCompare* ptestRWcompare)
+{
+    UINT countErrors = 0;
+
+    TestRedrawWindow(ptestRW);
+
+    if (ptestRW->resultRedraw)
+    {
+        if (ptestRWcompare->resultColorPre1 != ptestRW->resultColorPre1)
+        {
+            trace("aa");
+            trace("ERROR-resultColorPre1 %x %x\n", (int)ptestRW->resultColorPre1, (int)ptestRWcompare->resultColorPre1);
```

Comment:
Everywhere, `trace("ERROR-...` looks odd: use something like `ok(FALSE, ...)`?
Note: You may want to use a macro to pass the calling line number, or the like.

Reply:
I tried to write:
```
ok(ptestRW->resultColorPre1 != ptestRWcompare->resultColorPre1, "ERROR-resultColorPre1 %x %x\n",
			(int)ptestRW->resultColorPre1, (int)ptestRWcompare->resultColorPre1);
```
instead of:
`trace("ERROR-resultColorPre1 %x %x\n", (int)ptestRW->resultColorPre1, (int)ptestRWcompare->resultColorPre1);`
This would have the advantage of also outputting the line number of the problem, but that can easily be traced by the description and the line number of the overall error. What I don't like about it is that the number of tests would be non-concatenative and would add up both the individual variable checks and the test errors.

Reply:
Well, incrementing test error number on each check failure looks fine (as wanted) to me : why mask them and "need" extra calling `ok()`s and code and string bloats?
What's wrong with simple `ok(ptestRWcompare->resultColorPre1 == ptestRW->resultColorPre1, ...)` for example?

Then, changing strings to 'trace("DIFFERENCE' while keeping 'countErrors' looks even odder to me.

Reply:
1. i didn't want the total number of tests to be too high 36*X, but only 36.
2. If I put ok(...) only in the subprocedure, it would not be clear from the line number which test the error is related to.
I have no problem with removing trace("DIFFERENCE.

Reply:
1. i didn't want the total number of tests to be too high 36*X, but only 36.
2. If I put ok(...) only in the subprocedure, it would not be clear from the line number which test the error is related to.

The trace("DIFFERENCE helps me significantly to find the exact location of the error.

Reply:
It seems we have opposite views on how a test should be coded and behave.
As an additional reminder, `trace()` output is disabled on some testing bots, if I remember correctly...